piecrust/src/store/diff.rs

```diff
@@ -10,40 +10,72 @@ use std::io::{Read, Write};
 use qbsdiff::bsdiff::Bsdiff;
 use qbsdiff::bspatch::Bspatch;
 
-use crate::store::mmap::MmapMut;
+use crate::store::memory::Memory;
 
-/// Compute the diff between an `old` and a `new` buffer, and write it to thee
-/// given writer. The length of new buffer is written first, in the form of a
-/// `u64`.
+/// Compute the diff between the old and the new memory, and write it to the
+/// given writer.
+///
+/// First, the difference in size between the old and the new memory will be
+/// written, to handle growth of the memory when [`patch`] is called, and then
+/// the old memory will be grown to match the size of the new memory.
+///
+/// # Important
+/// The old memory should be discarded (dropped) after this operation.
 pub fn diff<T: Write>(
-    old: &[u8],
-    new: &[u8],
+    old: &Memory,
+    new: &Memory,
     writer: &mut T,
-) -> io::Result<u64> {
-    let new_len = new.len() as u64;
-    let new_len_bytes = new_len.to_le_bytes();
+) -> io::Result<()> {
+    let mut old = old.write();
+    let new = new.read();
 
-    writer.write_all(&new_len_bytes)?;
+    let old_len = old.len();
+    let new_len = new.len();
 
-    Bsdiff::new(old, new).compare(writer)
+    debug_assert!(
+        new_len >= old_len,
+        "Length of memories should strictly increase"
+    );
+
+    let delta_len = new_len - old_len;
+    let delta_len_bytes = delta_len.to_le_bytes();
+
+    writer.write_all(&delta_len_bytes)?;
+
+    if delta_len > 0 {
+        old.grow_by(delta_len)?;
+    }
+
+    Bsdiff::new(old.as_bytes(), new.as_bytes()).compare(writer)?;
+
+    Ok(())
 }
 
-/// Patches the given `mmap` with the given `patch` on top of the `old`
-/// buffer. The mmap will be grown if the length included in the diff is
-/// larger than it's current length.
+/// Applies the given patch on top of the new memory, using the old one as a
+/// base for the patch.
+///
+/// First, the growth delta is read from the patch, and if it's larger than zero
+/// both memories are grown. Then the patch is applied on top on the new memory,
+/// stemming from the contents of the old.
+///
+/// # Important
+/// The old memory should be discarded (dropped) after this operation.
```

Comment:
If it's so important, did you consider to take the ownership of the old Memory?

Reply:
Omg, you're right :man_facepalming:

Reply:
Well partially. The memory is just an `Arc` to an inner structure. This `Arc` should be deleted, and not used again.

Reply:
Damn, I was almost there